kbfsgit/runner.go

```diff
@@ -232,56 +267,184 @@ func (r *runner) waitForJournal(ctx context.Context) error {
 	return nil
 }
 
-func (r *runner) handleList(ctx context.Context, args []string) error {
+// handleList: From https://git-scm.com/docs/git-remote-helpers
+//
+// Lists the refs, one per line, in the format "<value> <name> [<attr>
+// …​]". The value may be a hex sha1 hash, "@<dest>" for a symref, or
+// "?" to indicate that the helper could not get the value of the
+// ref. A space-separated list of attributes follows the name;
+// unrecognized attributes are ignored. The list ends with a blank
+// line.
+func (r *runner) handleList(ctx context.Context, args []string) (err error) {
 	if len(args) > 0 {
 		return errors.New("Lists for non-fetches unsupported for now")
 	}
 
-	_, err := r.initRepoIfNeeded(ctx)
+	repo, err := r.initRepoIfNeeded(ctx)
 	if err != nil {
 		return err
 	}
 
+	refs, err := repo.References()
+	if err != nil {
+		return err
+	}
+
+	for {
+		ref, err := refs.Next()
+		if errors.Cause(err) == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		value := ""
+		switch ref.Type() {
+		case plumbing.HashReference:
+			value = ref.Hash().String()
+		case plumbing.SymbolicReference:
+			value = "@" + ref.Target().String()
+		default:
+			value = "?"
+		}
+		refStr := value + " " + ref.Name().String() + "\n"
+		_, err = r.output.Write([]byte(refStr))
+		if err != nil {
+			return err
+		}
+	}
+
 	err = r.waitForJournal(ctx)
 	if err != nil {
 		return err
 	}
 	r.log.CDebugf(ctx, "Done waiting for journal")
 
-	// TODO(KBFS-2353/KBFS-2354): list actual references from the repo.
 	_, err = r.output.Write([]byte("\n"))
 	return err
 }
 
-func (r *runner) processCommands(ctx context.Context) error {
+// handleFetchBatch: From https://git-scm.com/docs/git-remote-helpers
+//
+// fetch <sha1> <name>
+// Fetches the given object, writing the necessary objects to the
+// database. Fetch commands are sent in a batch, one per line,
+// terminated with a blank line. Outputs a single blank line when all
+// fetch commands in the same batch are complete. Only objects which
+// were reported in the output of list with a sha1 may be fetched this
+// way.
+//
+// Optionally may output a lock <file> line indicating a file under
+// GIT_DIR/objects/pack which is keeping a pack until refs can be
+// suitably updated.
+func (r *runner) handleFetchBatch(ctx context.Context, args [][]string) (
+	err error) {
+	repo, err := r.initRepoIfNeeded(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.log.CDebugf(ctx, "Fetching %d refs into %s", len(args), r.gitDir)
+
+	remote, err := repo.CreateRemote(&gogitcfg.RemoteConfig{
+		Name: localRepoRemoteName,
+		URL:  r.gitDir,
+	})
+
+	for _, fetch := range args {
+		if len(fetch) != 2 {
+			return errors.Errorf("Bad fetch request: %v", fetch)
+		}
+		refInBareRepo := fetch[1]
```

Comment:
This is the `<sha1>` in the doc above right?

Reply:
No, it's the `<name>`.